documentation/sphinx/source/developer-guide.rst

```diff
@@ -823,3 +823,133 @@ Loading data is a common task in any database. Loading data in FoundationDB will
+
+How Versions are Generated and assigned
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Versions are generated by the process that runs the *master* role. FoundationDB guarantees that no version will be generated twice and that the versions are monotonically increasing.
+
+In order to assign read and write versions to transactions, a client will never talk to a master. Instead it will get both from a proxy. Getting a read version is more complex than a commit version. Let's first look at commit versions:
+
+1. The client will send a commit message to a proxy (we will later look into how such a message looks like).
+1. The proxy will put this commit message in a queue in order to build a batch.
+1. In parallel, the proxy will ask for a new version from the master (note that this means that only proxies will ever ask for new versions - which scales much better as it puts less stress on the network).
+1. The proxy will then resolve all transactions withing that batch (discussed later) and assign the version it got from the master to *all* transactions within that batch. It will then write the transactions to the transaction log system to make it durable.
+1. If the transaction succeeded, it will send back the version as commit version to the client. Otherwise it will send back an error.
+
+As mentioned before, the algorithm to assign read versions is a bit more complex. At the start of a transaction, a client will ask a proxy server for a read version. The proxy will reply with the last committed version as of the time it received the request - this is important to guarantee external consistency. This is how this is achieved:
+
+#. The client will send a GRV (get read version) request to a proxy.
+#. The proxy will batch GRV requests for a short amount of time (it depends on load and configuartion how big these batches will be).
+#. The proxy will do the following steps in parallel:
+   * Ask all other proxies for their most recent committed version (the largest version they received from the master for which it successfully wrote the transactions to the transaction log system).
+   * Send a message to the transaction log system to verify that it is still writable. This is to prevent that we fetch read versions from a proxy that has been declared to be dead.
+#. It will then take the largest committed version from all proxies (including its own) and send it back to the clients.
+
+Conflict Detection
+~~~~~~~~~~~~~~~~~~
+
+This section will only explain conceptually how transactions are resolved in FoundationDB. The implementation will use multiple servers running the *Resolver* role and the keyspace will be sharded across them. It will also only allow to resolve transactions where the commit version will be larger than ``commitVersion - 5Million`` (around 5 seconds...).
```

Comment:
I wonder how the keyspace can be sharded? does the resolver ask other resolvers for historic writes to keys it has in its tx?

Reply:
Each resolver acts independently. The proxy combines the results from the resolvers and aborts the commit if any of them report a conflict. Each resolver conservatively assumes that if its key range had no conflicts then the transaction committed. This means that later on you might conflict with a transaction that didn't actually commit. This is legal because it's always legal for proxy to decide not to commit your transaction.

I thought this was documented somewhere but I couldn't find it /shrug

Reply:
so what is the sharding defined on? do they all have the full list of all txs - I hope not. Which means the space of keys or transactions or something needs to be partitioned.

Reply:
The keyspace is sharded - every transaction needs to be sent to every resolver.